crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -945,72 +945,65 @@ impl StickyData for SlidingSyncStickyParameters {
 // NOTE: SS proxy workaround.
 #[instrument(skip_all)]
 fn compute_limited(
-    known_rooms: &BTreeMap<OwnedRoomId, SlidingSyncRoom>,
-    response_rooms: &mut BTreeMap<OwnedRoomId, v4::SlidingSyncRoom>,
+    local_rooms: &BTreeMap<OwnedRoomId, SlidingSyncRoom>,
+    remote_rooms: &mut BTreeMap<OwnedRoomId, v4::SlidingSyncRoom>,
 ) {
-    for (room_id, room) in response_rooms {
+    for (room_id, remote_room) in remote_rooms {
         // Only rooms marked as initially loaded are subject to the fixup.
-        let initial = room.initial.unwrap_or(false);
+        let initial = remote_room.initial.unwrap_or(false);
         if !initial {
             continue;
         }
 
-        if room.limited {
+        if remote_room.limited {
             // If the room was already marked as limited, the server knew more than we do.
             continue;
         }
 
-        trace!(?room_id, "starting");
+        let remote_events = &remote_room.timeline;
+        if remote_events.is_empty() {
+            trace!("no timeline updates in the response => not limited");
+            continue;
+        }
 
         // If the known room had some timeline events, consider it's a `limited` if
         // there's absolutely no overlap between the known events and
         // the new events in the timeline.
-        if let Some(known_room) = known_rooms.get(room_id) {
-            let known_events = known_room.timeline_queue();
-
-            if room.timeline.is_empty() && !known_events.is_empty() {
-                // If the cached timeline had events, but the one in the response didn't have
-                // any, don't mark the room as limited.
-                trace!(
-                    "no timeline updates in response, local timeline had events => not limited."
-                );
+        if let Some(known_room) = local_rooms.get(room_id) {
+            let local_events = known_room.timeline_queue();
+
+            if local_events.is_empty() {
+                trace!("local timeline had no events => not limited");
                 continue;
             }
 
             // Gather all the known event IDs. Ignore events that don't have an event ID.
-            let num_known_events = known_events.len();
-            let known_events: HashSet<OwnedEventId> =
-                HashSet::from_iter(known_events.into_iter().filter_map(|event| event.event_id()));
-
-            if num_known_events != known_events.len() {
-                trace!(
-                    "{} local timeline events had no IDs",
-                    num_known_events - known_events.len()
-                );
-            }
+            let num_local_events = local_events.len();
+            let local_events_with_ids: HashSet<OwnedEventId> =
+                HashSet::from_iter(local_events.into_iter().filter_map(|event| event.event_id()));
```

Comment:
Slightly wasteful to do this if there are no remote events, because in that case we don't use this hashset anyways.

Reply:
If there are no remote events, we should've already skipped this at line 968. Or do you mean only remote events without IDs?

Reply:
Oh, it's further up & less indented. Nevermind. But, slightly related: Should the `if let` above be a `let else`?

Reply:
YASSS LESS INDENT :heart